V624 issue 10131

smuzaffar opened the pull request.

pcanal committed:
rootpcm: Read enums before class to get the correct underlying type.

pcanal committed:
When loading a TClass for map, also update the pair's TClass.

This is done also for multimap.

pcanal committed:
TStreamerInfo::Clear, also reset TStreamerElement's offset

pcanal committed:
rootcling: record Enum used by classes and collection.

When encountering a data member of a class that is a enum or if it is use in a map or multimap,
record the enum in the rootpcm.

pcanal committed:
io: Correct offset passed from SyntheticPair to StreamerInfo.

This was a fatal typo in 96784051cc6 pair: properly update StreamerInfo.

pcanal committed:
Add missing TListOfEnums::FindObject.

We already had TListOfEnumsWithLock but TListOfEnums::FindObject is used by rootcling

pcanal committed:
TGenCollectionProxy: add missing nullptr protection

pcanal committed:
When resetting a pair's StreamerInfo inform all collection proxies

pcanal committed:
pair: Also treat unordered map

pcanal committed:
TClass: clear fLastReadInfo, fCurrentInfo when deleting TStreamerInfos

pcanal committed:
Do the pair TClass StreamerInfo refresh only when needed

pcanal committed:
TProtoClass::FillTClass: Create a ListEnumsWithLock when needed

pcanal committed:
Don't put class' enum in rootpcm global list of enums

smuzaffar commented:
please test

cmsbuild commented:
A new Pull Request was created by  @smuzaffar (Malik Shahzad Muzaffar) for branch cms/v6-24-00-patches/4b08829.

@smuzaffar, @iarspider can you please review it and eventually sign? Thanks.
@perrotta, @dpiparo, @qliphy you are the release manager for this.
cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8c9a64/23542/summary.html
**COMMIT**: 6061468d166edbc80e35eba1a424c7394370dc1a
**CMSSW**: CMSSW_12_4_X_2022-03-30-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/root/167/23542/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3585896
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3585866
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 45 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-03-30-1100+8c9a64/49309/triggerResults): no differences found

pcanal committed:
TBufferFile: create StreamerInfo only if needed.

In the 'root_serialization' use case, we ended in TBufferFile::WriteClassBuffer with no
StreamerInfo created yet for the TClass for `std::vector<float>` and thus go into
the branch that creates the StreamerInfo ... However, at least in that case,
the call to `BuildRealData` induced the creation of the StreamerInfo but the code
in `TBufferFile::WriteClassBuffer` did not notice and thus create a second
one and tried to register it in the same slot.  (The actual negative side
effect was only an error message and a memory leak).

pcanal committed:
TGenCollectionProxy init: fail if the value's StreamerInfo is not there.

This happened in a user case where the container (map) for a pair of an enum, which had not dictionary
and something else that did have a dictionary.

pcanal committed:
Avoid reading pass the end of the StreamerInfo container.

This was happening when, inadvertently, StreamerInfo were created multiple times and then immeditately
deleted (for map<enumtype_without_dict, somethinelse>)

pcanal committed:
CollectionProxy: pass pair hint to R__CreateValue

cmsbuild commented:
Pull request #167 was updated.

pcanal committed:
Remove unnecessary trailing semi-colon

pcanal committed:
[NFC] Simplify code in TListOfEnums

cmsbuild commented:
Pull request #167 was updated.
